stratum/hal/lib/barefoot/p4runtime_bfrt_translator.h

```diff
@@ -0,0 +1,165 @@
+      SHARED_LOCKS_REQUIRED(lock_);
+  virtual bool TranslationEnabled() LOCKS_EXCLUDED(lock_);
+
+ protected:
+  // Default constructor.
+  P4RuntimeBfrtTranslator()
+      : device_id_(0),
+        bf_sde_interface_(nullptr),
+        translation_enabled_(false) {}
+
+ private:
+  // Private constructor. Use CreateInstance() to create an instance of this
+  // class.
+  P4RuntimeBfrtTranslator(BfSdeInterface* bf_sde_interface, int device_id)
+      : device_id_(device_id),
+        bf_sde_interface_(bf_sde_interface),
+        translation_enabled_(false) {}
+
+  // Reader-writer lock used to protect access to specific states.
+  mutable absl::Mutex lock_;
+
+  // Fixed zero-based BFRT device_id number corresponding to the node/ASIC
+  // managed by this class instance. Assigned in the class constructor.
+  const int device_id_;
+
+  // Pointer to a BfSdeInterface implementation that wraps all the SDE calls.
+  // Not owned by this class.
+  BfSdeInterface* bf_sde_interface_ = nullptr;
+
+  // Maps between singleton port and SDK port, vice versa
+  std::map<uint32, uint32> singleton_port_to_sdk_port_ GUARDED_BY(lock_);
+  std::map<uint32, uint32> sdk_port_to_singleton_port_ GUARDED_BY(lock_);
+  bool translation_enabled_ GUARDED_BY(lock_);
+
+  // P4Runtime translation information
+  std::map<uint32, std::map<uint32, std::string>> table_to_field_to_type_uri_
+      GUARDED_BY(lock_);
+  std::map<uint32, std::map<uint32, std::string>> action_to_param_to_type_uri_
+      GUARDED_BY(lock_);
+  std::map<uint32, std::map<uint32, std::string>> ctrl_hdr_to_meta_to_type_uri_
+      GUARDED_BY(lock_);
+  std::map<uint32, std::string> counter_to_type_uri_ GUARDED_BY(lock_);
+  std::map<uint32, std::string> meter_to_type_uri_ GUARDED_BY(lock_);
+  std::map<uint32, std::string> register_to_type_uri_ GUARDED_BY(lock_);
+
+  std::map<uint32, std::map<uint32, int32>> table_to_field_to_bit_width_
+      GUARDED_BY(lock_);
+  std::map<uint32, std::map<uint32, int32>> action_to_param_to_bit_width_
+      GUARDED_BY(lock_);
+  std::map<uint32, std::map<uint32, int32>> ctrl_hdr_to_meta_to_bit_width_
+      GUARDED_BY(lock_);
+  std::map<uint32, int32> counter_to_bit_width_ GUARDED_BY(lock_);
+  std::map<uint32, int32> meter_to_bit_width_ GUARDED_BY(lock_);
+  std::map<uint32, int32> register_to_bit_width_ GUARDED_BY(lock_);
+
+  friend class P4RuntimeBfrtTranslatorTest;
+};
+
+class P4RuntimeBfrtTranslationWriterWrapper
+    : public WriterInterface<::p4::v1::ReadResponse> {
+ public:
+  bool Write(const ::p4::v1::ReadResponse& msg) override;
+  ~P4RuntimeBfrtTranslationWriterWrapper() override {}
+  static std::unique_ptr<P4RuntimeBfrtTranslationWriterWrapper> CreateInstance(
+      WriterInterface<::p4::v1::ReadResponse>* writer,
+      P4RuntimeBfrtTranslator* p4runtime_bfrt_translator) {
+    return absl::WrapUnique(new P4RuntimeBfrtTranslationWriterWrapper(
+        writer, p4runtime_bfrt_translator));
+  }
```

Comment:
Since we're not doing any work here, we can probably use a normal contructor. See: `ServerWriterWrapper` and `ChannelWriterWrapper`. The managers generally have a `CreateInstance` for consistency, because _some_ do work.